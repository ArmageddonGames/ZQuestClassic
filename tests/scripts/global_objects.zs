#include "std.zh"

class Thing
{
	Thing otherThing;
	Counter someCounters[5];
}

class Counter
{
	int field;
}

Thing globalThing;

ffc script GlobalObject
{
	void run()
	{
		if (!globalThing)
		{
			globalThing = new Thing();
			globalThing->otherThing = new Thing();
			globalThing->someCounters[0] = new Counter();
			globalThing->otherThing->someCounters[0] = new Counter();
			GlobalObject(globalThing);
			GlobalObject(globalThing->otherThing);
			GlobalObject(globalThing->someCounters[0]);
			GlobalObject(globalThing->otherThing->someCounters[0]);
		}

		globalThing->someCounters[0]->field += 1;
		globalThing->otherThing->someCounters[0]->field -= 1;

		printf("globalThing->someCounters[0]->field %d\n", globalThing->someCounters[0]->field);
		printf("globalThing->otherThing->someCounters[0]->field %d\n", globalThing->otherThing->someCounters[0]->field);
	}
}
