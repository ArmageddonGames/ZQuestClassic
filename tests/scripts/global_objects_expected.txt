Compiling 'global_objects.zs'
Pass 1: Parsing
Pass 2: Preprocessing
Pass 3: Registration
Pass 4: Analyzing Code
Pass 5: Checking code paths
Pass 6: Generating object code
Pass 7: Assembling
Success!
Compile finished with exit code '0' (success)

PUSHR REFFFC; void run() Body Start
PUSHARGSV 0,0.0002
SETR d4,SP2
SETR d2,gd2; unless() #558 Test
COMPAREV d2,0
GOTOCMP 54,!=; Test 'unless'
SETV d2,0; ProtoFunc[Thing Thing()] Set Destructor AND unless() #558 Body Start
ZCLASS_CONSTRUCT d2,{0.0001,0.0005}
ZCLASS_MARK_TYPE {0,0.0001}; ProtoFunc[Thing Thing()] Default Construct
SET_OBJECT gd2,d2
SETV d2,0; ProtoFunc[Thing Thing()] Set Destructor
ZCLASS_CONSTRUCT d2,{0.0001,0.0005}
ZCLASS_MARK_TYPE {0,0.0001}; ProtoFunc[Thing Thing()] Default Construct
PUSHR d2
SETR d2,gd2
SETR d3,d2
POP d2
ZCLASS_WRITE d3,0
SETV d2,0; ProtoFunc[Counter Counter()] Set Destructor
ZCLASS_CONSTRUCT d2,{0.0001}; ProtoFunc[Counter Counter()] Default Construct
PUSHR d2
SETR d2,gd2
ZCLASS_READ d2,0.0001
SETR d0,d2
POP d2
WRITEPODARRAYVR 0,d2
SETV d2,0; ProtoFunc[Counter Counter()] Set Destructor
ZCLASS_CONSTRUCT d2,{0.0001}; ProtoFunc[Counter Counter()] Default Construct
PUSHR d2
SETR d2,gd2
ZCLASS_READ d2,0
ZCLASS_READ d2,0.0001
SETR d0,d2
POP d2
WRITEPODARRAYVR 0,d2
SETR d2,gd2; InlineFunc[void GlobalObject(untyped)] Params
ZCLASS_GLOBALIZE d2; InlineFunc[void GlobalObject(untyped)] Body
SETR d2,gd2; InlineFunc[void GlobalObject(untyped)] Params Start
ZCLASS_READ d2,0; InlineFunc[void GlobalObject(untyped)] Params End
ZCLASS_GLOBALIZE d2; InlineFunc[void GlobalObject(untyped)] Body
SETR d2,gd2; InlineFunc[void GlobalObject(untyped)] Params Start
ZCLASS_READ d2,0.0001
PUSHR d2
POP d0
READPODARRAYV d2,0; InlineFunc[void GlobalObject(untyped)] Params End
ZCLASS_GLOBALIZE d2; InlineFunc[void GlobalObject(untyped)] Body
SETR d2,gd2; InlineFunc[void GlobalObject(untyped)] Params Start
ZCLASS_READ d2,0
ZCLASS_READ d2,0.0001
PUSHR d2
POP d0
READPODARRAYV d2,0; InlineFunc[void GlobalObject(untyped)] Params End
ZCLASS_GLOBALIZE d2; InlineFunc[void GlobalObject(untyped)] Body AND unless() #558 Body End
SETR d2,gd2
ZCLASS_READ d2,0.0001
PUSHR d2
POP d0
READPODARRAYV d2,0
ZCLASS_READ d2,0
ADDV d2,1
PUSHR d2
SETR d2,gd2
ZCLASS_READ d2,0.0001
PUSHR d2
POP d0
READPODARRAYV d2,0
SETR d3,d2
POP d2
ZCLASS_WRITE d3,0
SETR d2,gd2
ZCLASS_READ d2,0
ZCLASS_READ d2,0.0001
PUSHR d2
POP d0
READPODARRAYV d2,0
ZCLASS_READ d2,0
SUBV d2,1
PUSHR d2
SETR d2,gd2
ZCLASS_READ d2,0
ZCLASS_READ d2,0.0001
PUSHR d2
POP d0
READPODARRAYV d2,0
SETR d3,d2
POP d2
ZCLASS_WRITE d3,0
ALLOCATEMEMV d2,40,0
STORE d2,0.0001
WRITEPODSTRING d2,"globalThing->someCounters[0]->field %d\n"
LOAD d2,0.0001; InlineFunc[void printf(char32)] Params Start
PUSHR d2
SETR d2,gd2
ZCLASS_READ d2,0.0001
PUSHR d2
POP d0
READPODARRAYV d2,0
ZCLASS_READ d2,0
PUSHVARGR d2; InlineFunc[void printf(char32)] Params End
PRINTFVARG; InlineFunc[void printf(char32)] Body Start
POP d5; InlineFunc[void printf(char32)] Body End
LOAD d3,0.0001
DEALLOCATEMEMR d3
ALLOCATEMEMV d2,52,0
STORE d2,0
WRITEPODSTRING d2,"globalThing->otherThing->someCounters[0]->field %d\n"
LOAD d2,0; InlineFunc[void printf(char32)] Params Start
PUSHR d2
SETR d2,gd2
ZCLASS_READ d2,0
ZCLASS_READ d2,0.0001
PUSHR d2
POP d0
READPODARRAYV d2,0
ZCLASS_READ d2,0
PUSHVARGR d2; InlineFunc[void printf(char32)] Params End
PRINTFVARG; InlineFunc[void printf(char32)] Body Start
POP d5; InlineFunc[void printf(char32)] Body End
LOAD d3,0
DEALLOCATEMEMR d3
POPARGS d5,0.0003
QUIT; void run() Body End


MARK_TYPE_REG gd2,0.0001; void run() Body Start
SETV gd2,0
ALLOCATEGMEMV d2,4096,0
SETR gd1,d2
QUIT; void run() Body End